beacon-chain/sync/rpc_beacon_blocks_by_range.go

```diff
@@ -22,97 +22,107 @@ func (r *Service) beaconBlocksByRangeRPCHandler(ctx context.Context, msg interfa
 
+	// ticker to stagger out large requests.
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
 	m, ok := msg.(*pb.BeaconBlocksByRangeRequest)
 	if !ok {
 		return errors.New("message is not type *pb.BeaconBlockByRangeRequest")
 	}
 
+	// the initial count for the first batch to be returned back.
+	count := m.Count
+	if count > allowedBlocksPerSecond {
+		count = allowedBlocksPerSecond
+	}
+	// initial batch parameters to be returned to remote peer.
 	startSlot := m.StartSlot
-	endSlot := startSlot + (m.Step * (m.Count - 1))
-	remainingBucketCapacity := r.blocksRateLimiter.Remaining(stream.Conn().RemotePeer().String())
+	endSlot := startSlot + (m.Step * (count - 1))
 
+	// final requested slot from remote peer.
+	endReqSlot := startSlot + (m.Step * (m.Count - 1))
+
+	remainingBucketCapacity := r.blocksRateLimiter.Remaining(stream.Conn().RemotePeer().String())
 	span.AddAttributes(
 		trace.Int64Attribute("start", int64(startSlot)),
-		trace.Int64Attribute("end", int64(endSlot)),
+		trace.Int64Attribute("end", int64(endReqSlot)),
 		trace.Int64Attribute("step", int64(m.Step)),
 		trace.Int64Attribute("count", int64(m.Count)),
 		trace.StringAttribute("peer", stream.Conn().RemotePeer().Pretty()),
 		trace.Int64Attribute("remaining_capacity", remainingBucketCapacity),
 	)
+	for startSlot <= endReqSlot {
+		remainingBucketCapacity = r.blocksRateLimiter.Remaining(stream.Conn().RemotePeer().String())
 
-	if m.Count > uint64(remainingBucketCapacity) {
-		r.p2p.Peers().IncrementBadResponses(stream.Conn().RemotePeer())
-		if r.p2p.Peers().IsBad(stream.Conn().RemotePeer()) {
-			log.Debug("Disconnecting bad peer")
-			defer func() {
-				if err := r.p2p.Disconnect(stream.Conn().RemotePeer()); err != nil {
-					log.WithError(err).Error("Failed to disconnect peer")
-				}
-			}()
-		}
-		resp, err := r.generateErrorResponse(responseCodeInvalidRequest, rateLimitedError)
-		if err != nil {
-			log.WithError(err).Error("Failed to generate a response error")
-		} else {
-			if _, err := stream.Write(resp); err != nil {
-				log.WithError(err).Errorf("Failed to write to stream")
+		if allowedBlocksPerSecond > uint64(remainingBucketCapacity) {
+			r.p2p.Peers().IncrementBadResponses(stream.Conn().RemotePeer())
+			if r.p2p.Peers().IsBad(stream.Conn().RemotePeer()) {
+				log.Debug("Disconnecting bad peer")
+				defer func() {
+					if err := r.p2p.Disconnect(stream.Conn().RemotePeer()); err != nil {
+						log.WithError(err).Error("Failed to disconnect peer")
+					}
+				}()
 			}
+			r.writeErrorResponseToStream(responseCodeInvalidRequest, rateLimitedError, stream)
+			return errors.New(rateLimitedError)
 		}
-		return errors.New(rateLimitedError)
-	}
+		r.blocksRateLimiter.Add(stream.Conn().RemotePeer().String(), int64(allowedBlocksPerSecond))
 
-	r.blocksRateLimiter.Add(stream.Conn().RemotePeer().String(), int64(m.Count))
+		// TODO(3147): Update this with reasonable constraints.
+		if endSlot-startSlot > 1000 || m.Step == 0 {
```

Comment:
what is this `1000` magic number? should we define it at the top and use a const?

Reply:
Not actually sure, we have had it for as long as I can remember
@prestonvanloon any idea why its 1000 ?

Reply:
There is no reason. I put 1000 because I felt that was a large range and anything more than that would be too much to fetch from the DB at once.